include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h

```diff
@@ -149,6 +149,33 @@ struct _ModeConverter<access_mode::write>
     static constexpr access_mode __value = access_mode::discard_write;
 };
 
+template <typename Iter, typename ValueType = typename std::iterator_traits<Iter>::value_type>
+using __default_alloc_vec_iter = typename std::vector<ValueType>::iterator;
+
+template <typename Iter, typename ValueType = typename std::iterator_traits<Iter>::value_type>
+using __usm_shared_alloc_vec_iter =
+    typename std::vector<ValueType, typename sycl::usm_allocator<ValueType, sycl::usm::alloc::shared>>::iterator;
+
+template <typename Iter, typename ValueType = typename std::iterator_traits<Iter>::value_type>
+using __usm_host_alloc_vec_iter =
+    typename std::vector<ValueType, typename sycl::usm_allocator<ValueType, sycl::usm::alloc::host>>::iterator;
+// Evaluates to true if the provided type is an iterator with a value_type and if the implementation of a
+// std::vector<value_type, Alloc>::iterator can be distinguished between three different allocators, the
+// default, usm_shared, and usm_host. If all are distinct, it is very unlikely any non-usm based allocator
+// could be confused with a usm allocator.
+template <typename Iter, typename Void = void>
+struct __vector_iter_distinguishes_by_allocator : std::false_type
+{
+};
+template <typename Iter>
+struct __vector_iter_distinguishes_by_allocator<
+    Iter, std::enable_if_t<!std::is_same_v<__default_alloc_vec_iter<Iter>, __usm_shared_alloc_vec_iter<Iter>> &&
```

Comment:
Should we apply here `std::decay_t` for `Iter` template param?

Reply:
Actually, this is already merged code from #1438 .  Working to rebase, to make this focus on just what is added by this PR.  
If we need to go back and fix something here, it will needs its own PR I think.